src/CTA.WebForms2Blazor/Services/TaskManagerService.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CTA.WebForms2Blazor.Services
+{
+    public class TaskManagerService
+    {
+        public enum TaskState { Active, Waiting }
+        public enum TaskStallTimeout { None = 0, Short = 100, Medium = 1000, Long = 10000 }
+
+        private IDictionary<int, ManagedTask> _managedTasks;
+        private bool _stalled;
+        private int _nextAvailableTaskId;
+
+        public TaskManagerService()
+        {
+            _managedTasks = new Dictionary<int, ManagedTask>();
+        }
+
+        public int RegisterNewTask()
+        {
+            var newManagedTask = new ManagedTask(_nextAvailableTaskId);
+            _nextAvailableTaskId += 1;
+
+            _managedTasks.Add(newManagedTask.TaskId, newManagedTask);
+            UpdateStallState();
+
+            return newManagedTask.TaskId;
+        }
+
+        public async Task<TResult> ManagedRun<TResult>(int taskId, Func<CancellationToken, Task<TResult>> func)
+        {
+            TResult result;
+            var managedTask = _managedTasks[taskId];
+            var token = managedTask.SetWaiting();
+            UpdateStallState();
+
+            try
+            {
+                result = await func(token);
+                managedTask.SetActive();
+            }
+            finally
+            {
+                UpdateStallState();
+            }
+
+            return result;
+        }
+
+        public void RetireTask(int taskId)
+        {
+            _managedTasks.Remove(taskId);
+            UpdateStallState();
+        }
+
+        private void UpdateStallState()
+        {
+            if (_managedTasks.Count > 0 && !_managedTasks.Values.Any(managedTask => managedTask.CurrentTaskState == TaskState.Active))
+            {
+                if (!_stalled)
+                {
+                    _stalled = true;
+                    TryResolveStall();
+                }
+            }
+            else
+            {
+                _stalled = false;
```

Comment:
Following up on what Jonathan said, I think instead of adding another else statement that sets _stalled to false, just move the the TryResolveStall() call outside so it gets called regardless and tries to fix it

Reply:
To answer both of you questions, we don't want to set stalled to false if the second if condition is not met, this is because the conditions for a stall have been met, but the state of the application was previously unstalled, if something about the tasks changes and the stall condition is still met, then we don't want to re-execute TryResolveStall() or set _stalled to false, we specifically want to do nothing in this case. Also @KallanBao the if statement within TryResolveStall serves a different purpose, see the reply to the comment below for that purpose